DM-21222: Stop forcing existence of credentials in connection strings.

DinoBektesevic opened the pull request:
Not forcing username and password in connection strings allows other mechanisms, such as Oracle Wallet, to insert the credentials. 

It was difficult to untangle `getAuth` into sub-functionality so `exists` is implemented in terms of `getAuth`. Is that ok?

Would this work?

DinoBektesevic committed:
Stop forcing existence of credentials in connection strings.

timj commented on python/lsst/daf/butler/core/connectionString.py:
```diff
@@ -80,12 +80,20 @@ def fromConfig(cls, registryConfig):
-            conStr.username = auth[0]
-            conStr.password = auth[1]
+        except DbAuthError as e:
+            if 'permissions' in e.args[0]:
```
Don't rely on parsing the error message in an attempt to work out what type of problem this was. What you need to do is raise a different exception for permissions problem. You want to be able to code this as something like:
```python
try:
    dbAuth = DbAuth(...)
except DBAuthFileNotFoundError:
   pass
else:
   ...
```
and have the constructor throw `DbAuthPermissionsError` that would not be caught. They could both inherit from `DbAuthError` but you'd only be ignoring the lack of db auth file rather than ignoring everything and trying to parse the error.

DinoBektesevic replied:
Yeah, I was pretty sure this would get a comment but wasn't sure whether a `DbAuthPermissionsError` is wanted ot should I just raise a regular python `PermissionError` (which isn't there exactly for this thing, but matched the closest to what I wanted).

timj replied:
DbAuthPermissionError could inherit from PermissionError and from DbAuthError depending on how you want people to be catching things.

ktlim commented:
I was just thinking of removing most of the error `raise`s, allowing `username` and `password` to be returned as empty strings (or `None`).

Reading through the file twice (for `exists` and then `getAuth`) seems undesirable.

DinoBektesevic committed:
Allow username or pasword to be None.

Instead of parsing through the file twice with `exists` and then
`getAuth` re-write the `DbAuth.getAuth` method to allow username
and password to be None.

This requires silencing any errors when credentials are
miss-configured from ConnectionStringFactory's side. Short of
having an explicit key in config stating whether a credentials
lookup is desired there is no clear cut way to say if we must
insist that username/password must exist or not.

Instead, we believe the provided db string and try to return it
unchanged as often as possible unless everything is configured
correctly.

DinoBektesevic commented:
Ok. redid it the way @ktlim wanted. 

Username and password are allowed to be `None.` Instead of avoiding `sqlite` specifically, we avoid any situation where `host` is not specified explicitly. `ConnectionStringFactory` now believes the correctness of given connection string optimistically.

brianv0 reviewed (approved):
Looks fine, any PR where you end up with less LOC is usually good 👍

DinoBektesevic commented:
@ktlim You are probably quite busy this week, but could you take a look and comment on the changes. Is this more like what you had in mind?

Someone should also run Jenkins on this because I don't have the privileges too.

timj reviewed (approved):
Looks okay to me. I'll do a Jenkins run. I don't think we need to wait for @ktlim to also approve.
